outlines/outline.py

```diff
@@ -0,0 +1,52 @@
+import ast
+from dataclasses import dataclass
+
+
+@dataclass
+class Outline:
+    """
+    Outline is a class that creates a callable object to generate responses
+    based on a given model and a prompt template.
+
+    Args:
+        model: The model to be used for generating responses.
+        template (function): A function that takes arguments and returns a prompt string.
+        output_type: The expected output type of the generated response.
+
+    Example:
+        from outlines import models
+
+        model = models.transformers("gpt2")
+
+        def template(a: int) -> str:
+            return f"What is 2 times {a}?"
+
+        fn = Outline(model, template, int)
+
+        result = fn(3)
+        print(result)  # Expected output: 6
+    """
+
+    def __init__(self, model, template, output_type):
+        self.model = model
+        self.template = template
+        self.output_type = output_type
+
+    def __call__(self, *args):
+        # Generate the prompt using the template function
+        prompt = self.template(*args)
+
+        # Generate the response using the model
+        response = self.model.generate(prompt)
```

Comment:
This should use either `outlines.generate.json`, `outlines.generate.regex`, etc. The interface is necessarily going to be brittle for now. You can take a look at the v1 branch to see how this is going to be more robust with the undergoing refactor.

Reply:
Hmm, okay, but which one of those should I use by default? And how should I let users switch to the other one, by adding a fourth optional argument, e.g. `generation_kind`, WDYT?

Reply:
I just realized that maybe I misunderstood the PR, and it should infer the correct regex to match the user's required output type. E.g., if a user asks for an `int`, I should ask `outlines.generate.regex` something like `^[+-]?[0-9]+$`. IIUC, I'm unsure how to scale this to arbitrary return types. Should we define a list of supported return types? Using `ast.literal_eval` was the most flexible solution I came up with, but I'm not sure how well it handles custom data structures. In such cases, I think we should check that users provide types that implement deserialization from JSON. WDYT?

Reply:
Let's only accept Pydantic models / JSON Schema strings for now and use `outlines.generate.json`. We'll be able to easily generalise after we push the refactor in the v1 branch.

Reply:
I've updated the PR _(and the usage example in the docstring)_. Does it still look like what we want?